cpp/katz_centrality_module/algorithm/katz.cpp

```diff
@@ -0,0 +1,405 @@
+    context_new.omegas[i][v] -= katz_alg::context.omegas[i - 1][w];
+  }
+
+  // Update centralities as last step
+  for (auto w : updated_nodes) {
+    if (i != 1) {
+      katz_alg::context.centralities[i][w] =
+          katz_alg::context.centralities[i - 1][w] + pow(katz_alg::alpha, i) * context_new.omegas[i][w];
+    } else {
+      katz_alg::context.centralities[i][w] +=
+          pow(katz_alg::alpha, i) * (context_new.omegas[i][w] - katz_alg::context.omegas[i][w]);
+    }
+  }
+}
+
+std::vector<std::pair<std::uint64_t, double>> WrapResults() {
+  return std::vector<std::pair<std::uint64_t, double>>(
+      katz_alg::context.centralities[katz_alg::context.iteration].begin(),
+      katz_alg::context.centralities[katz_alg::context.iteration].end());
+}
+
+bool IsIncosistent(const mg_graph::GraphView<> &graph) {
+  for (auto const [node_id] : graph.Nodes()) {
+    auto external_id = graph.GetMemgraphNodeId(node_id);
+    if (katz_alg::context.centralities[katz_alg::context.iteration].find(external_id) ==
+        katz_alg::context.centralities[katz_alg::context.iteration].end()) {
+      return true;
+    }
+  }
+  return false;
+}
+}  // namespace
+
+std::vector<std::pair<std::uint64_t, double>> GetKatz(const mg_graph::GraphView<> &graph) {
+  // Create context and calculate values if not initialized
+  if (!katz_alg::context.IsInitialized()) {
+    return SetKatz(graph);
+  }
+
+  if (IsIncosistent(graph)) {
+    throw std::runtime_error(
+        "Graph has been modified, therefore is incosistent with cached results, please update the Katz centrality by "
+        "calling set/reset!");
+  }
+
+  return WrapResults();
+}
+
+std::vector<std::pair<std::uint64_t, double>> SetKatz(const mg_graph::GraphView<> &graph, double alpha,
+                                                      double epsilon) {
+  katz_alg::alpha = alpha;
+  katz_alg::k = k;
+  katz_alg::epsilon = epsilon;
+  katz_alg::context.Init(graph);
+
+  if (graph.Edges().empty()) {
+    return WrapResults();
+  }
+
+  auto deg_max = MaxDegree(graph);
+  double gamma = deg_max / (1. - (alpha * alpha * deg_max));
+
+  // Initialize the active vector
+  std::transform(graph.Nodes().begin(), graph.Nodes().end(),
+                 std::inserter(katz_alg::context.active_nodes, katz_alg::context.active_nodes.end()),
+                 [&graph](mg_graph::Node<> vertex) -> std::uint64_t { return graph.GetMemgraphNodeId(vertex.id); });
+
+  return KatzCentralityLoop(katz_alg::context.active_nodes, graph, katz_alg::alpha, katz_alg::k, katz_alg::epsilon,
+                            gamma);
+}
+
+std::vector<std::pair<std::uint64_t, double>> UpdateKatz(
+    const mg_graph::GraphView<> &graph, const std::vector<std::uint64_t> &new_vertices,
+    const std::vector<std::pair<std::uint64_t, uint64_t>> &new_edges, const std::vector<std::uint64_t> &new_edge_ids,
+    const std::vector<std::uint64_t> &deleted_vertices,
+    const std::vector<std::pair<std::uint64_t, uint64_t>> &deleted_edges) {
```

Comment:
According to the paper, after a dynamic update one needs to check for convergence and if the algorithm doesn't converge anymore, run iterations of the static algorithm until it does. This step appears to be missing, did you find an alternative?

Reply:
Static algorithm convergence is done in the method called `KatzCentralityLoop`

Reply:
It's applied after initial handling of dynamic case.

Reply:
I see, thank you!